### ISSUE TITLE
Symbolic Link permission when compiled with musl libc on Alpine Linux

### ISSUE THREAD

**Issue by Soldering-Stan** [RsyncProject/rsync#132]
When using rsync to copy from a *bsd host to Linux when rsync is compiled against musl libc, rsync issues warnings during the copy :-

rsync: failed to set permissions on "FILENAME": Not supported (95)

and exits with error code 23 :- 

```
HOSTNAME:/tmp/testing# rsync -va SHOST:/etc .
receiving incremental file list
rsync: failed to set permissions on "/tmp/testing/etc/aliases": Not supported (95)
rsync: failed to set permissions on "/tmp/testing/etc/rmt": Not supported (95)
rsync: failed to set permissions on "/tmp/testing/etc/termcap": Not supported (95)
rsync: failed to set permissions on "/tmp/testing/etc/unbound": Not supported (95)
rsync: failed to set permissions on "/tmp/testing/etc/ssl/cert.pem": Not supported (95)

sent 71 bytes  received 8,713 bytes  5,856.00 bytes/sec
total size is 1,689,030  speedup is 192.28
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1659) [generator=3.1.3]
HOSTNAME:/tmp/testing# echo $?
23

```

These are all symbolic links.  On the source host (a freebsd box) the symlinks have the following permissions :-

```
lrwxr-xr-x   1 root  wheel        12 Jul 21  2017 aliases -> mail/aliases
lrwxr-xr-x   1 root  wheel        15 Jul 21  2017 rmt -> ../usr/sbin/rmt
lrwxr-xr-x   1 root  wheel        23 Jul 21  2017 termcap -> /usr/share/misc/termcap
lrwxr-xr-x   1 root  wheel        14 Jul 21  2017 unbound -> ../var/unbound
``` 
As Linux has no idea of anything other than lrwxrwxrwx permissions for symbolic links, can an option be added to rsync (or even an #ifdef #endif section along the lines of 'quietly ignore not supported error if linux && symlink;' ?

rsync works fine using GNU libc, it's just against musl.  There are several articles around this regarding musl but the consensus seems to be that it's not really a musl issue, rather it's a GNU libc quirk that accepts the fact that an lchmod just won't work on Linux.

**Comment by agostonbarna** [comment]
The symbolic link permission handling is also problematic when rsync is used from a Mac to a Linux machine. Symlinks have 755 permission on the Mac, but they are created with 777 on the Linux machine as it does not support changing symlink permissions.

As symlink permission doesn't matter, this shouldn't be a problem, but it is because on subsequent runs rsync compares the symlink permissions and even if they haven't changed it thinks they differ due to the different permissions, so it copies the links every time.

Would it be possible for rsync to ignore the symlink permissions if the source system (e.g. BSD, Mac) supports setting symlink permissions, but the target (e.g. Linux) does not, or provide a CLI flag to explicitly disable the symlink permission checking/copying of rsync (but still enable it for other files/dirs with the --perms flag)?

**Comment by WayneD** [comment]
This is something that we definitely need to improve a bit more, but it's rather complicated since the lack of symlink permission setting often depends on the destination filesystem, not the OS that compiled rsync.  So, rsync doesn't know if it can or cannot set a symlink's permissions until it tries to set its permissions.  We probably need some kind of per-device symlink-permission test method and have the code notice if the directory dev_no values are changing at mount boundaries where a new test would be needed.

Note, however, that rsync listing the symlink name for a transfer doesn't mean that it actually made any changes.  Hopefully you are using `-i` (I'm surprised that a lot of people do not) so that it shows you that it's a difference of permissions, which can be easily ignored or filtered out.

As for the "Not  supported" errors, I believe I already fixed that in a commit that should have been released in 3.2.4.

**Comment by agostonbarna** [comment]
Hey @WayneD
I'm glad to hear that the symlink handling might be improved in the future.

I'm using rsync with the -i flag and it shows the difference in the symlink permissions on each run, and my impression was it also copies the symlinks again, but then I just misinterpreted it. Filtering out the L...p from the rsync output is a good enough workaround for me.
Thanks for the suggestion!

**Comment by devZer0** [comment]
does this issue still exist?

according to https://github.com/RsyncProject/rsync/commit/85b8dc8abaca96fc3ea7421e09101b6ac41b6718 there should be no lchmod usage on linux anymore

**Comment by BraMo-logicar** [comment]
I have a very similar case, but at a much larger scale, where this may also have a significant performance impact.

I'm currently using rsync for a large filesystem migration from **BeeGFS to NFSv4**, involving about **80 TiB of data and 105,603,622 filesystem entries**.

BeeGFS reports symbolic links with mode `0755`, while the NFSv4 destination reports them as `0777`.

I routinely use `-i` during these migrations. On subsequent rsync runs, unchanged symlinks are reported as:

```
.L...p.....
```

In one of the areas I'm currently migrating, I have **1,035,112 symlinks** showing this permission difference on every run.

So I have two questions/suggestions.

### 1. What is the actual cost of this permission difference?

I understand that `.L...p.....` means that rsync detects a permission difference, but I'm not sure whether this only affects comparison/itemization or whether rsync subsequently attempts a `chmod`-like operation on each of these symlinks.

This distinction matters quite a lot with more than one million symlinks. This particular area is significantly slower to synchronize than others, so I'm trying to determine whether repeatedly processing these permission differences could be a relevant part of the cost.

### 2. Would an `--omit-link-perms` option make sense?

There is already:

```
--omit-link-times, -J    omit symlinks from --times
```

An analogous option could make `--perms` ignore permission differences for symlinks, while still preserving permissions for regular files and directories.

In other words:

```
--perms --omit-link-perms
```

would preserve normal `--perms` semantics for files and directories, but would neither compare nor attempt to update the permission bits of symlinks.

`--no-perms` is not a suitable workaround in this case because preserving file and directory permissions is required.

This would solve the BeeGFS -> NFSv4 case without special-casing either filesystem, and might also address the original FreeBSD -> Linux use case described in this issue.

**Comment by steadytao** [comment]
The original musl/Linux ENOTSUP error was fixed in 3.2.4 by disabling lchmod on Linux.

The current `.L...p.....` report is a separate remaining issue. Rsync still compares the symlink mode but on Linux `do_chmod()` returns without making a chmod system call so the link is not retransferred or repeatedly modified. There may still be meaningful comparison and itemisation overhead across ~ one million links.

An `--omit-link-perms` option analogous to `--omit-link-times` seems reasonable. Please provide the exact rsync version, command and timings with and without itemised output so the remaining cost can be measured.

**Comment by BraMo-logicar** [comment]
Thanks for the feedback.

Since the real workload contains millions of filesystem entries and is continuously modified by users, I need some time to prepare a controlled and reproducible benchmark.

I'm currently isolating a representative static subset with a very high symlink density, while preserving the original BeeGFS → NFSv4 architecture and, as far as possible, the actual metadata differences between the source and destination.

My first step will be to characterize the behaviour of stock rsync under identical conditions, comparing runs without itemization, with -i, and with -ii. Once I have reproducible results, I'll be able to investigate the symlink permission behaviour separately.

Since my case is somewhat different from the original issue — RHEL/glibc, BeeGFS → NFSv4, and potentially a performance impact at large scale — do you think it would be better to open a separate issue for the results and cross-reference #132, rather than continue the investigation here?

**Comment by steadytao** [comment]
Hmmm. I do think that would be best. We can likely close this one off. Thank you for your work 🙂‍↕️

**Comment by BraMo-logicar** [comment]
Sounds good, thank you.

I'll proceed with the controlled benchmark and open a separate issue once I have reproducible results, referencing this one for context.

Thanks again for your help!
